ledger/participant-integration-api/src/test/suite/scala/platform/store/appendonlydao/events/ACSReaderSpec.scala

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.appendonlydao.events
+
+import akka.actor.ActorSystem
+import akka.stream.Materializer
+import akka.stream.scaladsl.Sink
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.Duration
+
+class ACSReaderSpec extends AsyncFlatSpec with Matchers with BeforeAndAfterAll {
+
+  private val actorSystem = ActorSystem()
+  private implicit val materializer: Materializer = Materializer(actorSystem)
+  private implicit val ec: ExecutionContext = actorSystem.dispatcher
+
+  override def afterAll(): Unit = {
+    Await.result(actorSystem.terminate(), Duration(10, "seconds"))
+    ()
+  }
+
+  behavior of "pullWorkerSource"
+
+  it should "give an empty source if initialTasks are empty" in {
```

Comment:
Please add one test with parallelism 2. There will be new class of issues when the code is used concurently.

Reply:
oh boy, I saw that coming :(
actually doing parallel tests with expectations is not something which is easy (we need to hand-orchestrate completion from test suit. time based expectations just not work in our CI).
I will figure something out, since this is a workhorse of the 2 phase retrieval :+1:

Reply:
Added some more elaborate tests for increased parallelism, even a puppetmaster test where asynchronous behavior is analyzed in great detail